llama_index/llama_dataset/base.py

```diff
@@ -206,21 +232,65 @@ async def _apredict_example(
+        """Batches examples and predictions with a given batch_size."""
+        num_examples = len(self.examples)
+        for ndx in range(start_position, num_examples, batch_size):
+            yield self.examples[ndx : min(ndx + batch_size, num_examples)]
+
     async def amake_predictions_with(
-        self, query_engine: BaseQueryEngine, show_progress: bool = False
+        self,
+        query_engine: BaseQueryEngine,
+        show_progress: bool = False,
+        batch_size: int = 20,
+        sleep_time_in_seconds: int = 10,
     ) -> BaseLlamaPredictionDataset:
         """Async predict with a given query engine.
 
         Args:
             query_engine (BaseQueryEngine): The query engine to make predictions with.
             show_progress (bool, optional): Show progress of making predictions.
+            batch_size (int): Used to batch async calls, especially to reduce chances
+                            of hitting RateLimitError from openai.
+            sleep_time_in_seconds (int): Amount of time to sleep between batch call
+                            to reduce chance of hitting RateLimitError from openai.
 
         Returns:
             BaseLlamaPredictionDataset: A dataset of predictions.
         """
-        tasks = []
-        for example in self.examples:
-            tasks.append(self._apredict_example(query_engine, example))
-        asyncio_mod = asyncio_module(show_progress=show_progress)
-        predictions = await asyncio_mod.gather(*tasks)
-        return self._construct_prediction_dataset(predictions=predictions)
+        if self._predictions_cache:
+            start_example_position = len(self._predictions_cache)
+        else:
+            start_example_position = 0
+        print(f"start_example_position: {start_example_position}", flush=True)
+        for batch in self._batch_examples(
+            batch_size=batch_size, start_position=start_example_position
+        ):
+            tasks = []
+            for example in batch:
+                tasks.append(self._apredict_example(query_engine, example))
+            asyncio_mod = asyncio_module(show_progress=show_progress)
+
+            try:
+                batch_predictions = await asyncio_mod.gather(*tasks)
+            except RateLimitError as err:
+                print(f"num predictions cached: {len(self._predictions_cache)}")
+                raise ValueError(
+                    "You've hit rate limits on your OpenAI subscription. This"
+                    " class caches previous predictions after each successful"
+                    " batch execution. Based off this cache, when executing this"
+                    " command again it will attempt to predict on only the examples "
+                    "that have not yet been predicted. Try reducing your batch_size."
+                ) from err
+            self._predictions_cache += batch_predictions
+            time.sleep(sleep_time_in_seconds)
```

Comment:
should use `await asyncio.sleep()` ?

Reply:
Yes, thanks! I'm going to use asyncio instead :)